kickstarter/android-oss
# NT-1914: UX – Post new comment

# 📲 What

UX – Post new comment

# 🤔 Why

Implement the interaction responsible for posting a new comment.
- A user can post a comment

- Comment is post through GraphQL

- Comment is shown as the first of the list

- User is scrolled to the top if not there already

# 🛠 How

- Called createComment from commentViewModel
``` 

initialProject
                .compose(Transformers.takePairWhen(this.postComment))
                .compose(bindToLifecycle())
                .switchMap {
                    it.first?.let { project ->
                        this.apolloClient.createComment(
                            PostCommentData(
                                project = project,
                                body = it.second,
                                clientMutationId = null,
                                parentId = null
                            )
                        )
                    }
                }
                .subscribe(
                    {
                    },
                    {
                    }
                )

```

- Return posted comment from commentViewModel to be inserted into view
```
this.currentUser.loggedInUser()
                .compose(Transformers.takePairWhen(this.postComment))
                .compose(bindToLifecycle())
                .subscribe {
                    val comment = Comment.builder()
                        .body(it.second)
                        .parentId(-1)
                        .authorBadges(listOf())
                        .createdAt(DateTime.now())
                        .cursor("")
                        .deleted(false)
                        .id(-1)
                        .repliesCount(0)
                        .author(it.first)
                        .build()

                    this.insertComment.onNext(comment)
                }

```

- Inserted item into list and Scrolled list to top 
```
 viewModel.outputs.insertComment()
            .compose(bindToLifecycle())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe {
                adapter.insertData(it, 0)
                binding.commentsRecyclerView.scrollToPosition(0)
            }
```

- Added test
```
 @Test
    fun testCommentsViewModel_PostComment_CommentAddedToView() {
        val userAvatar = AvatarFactory.avatar()
        val currentUser = UserFactory.user().toBuilder().id(1).avatar(
            userAvatar
        ).build()

        val env = environment().toBuilder().apolloClient(object : MockApolloClient() {
            override fun createComment(comment: PostCommentData): Observable<Comment> {
                return Observable.just(CommentFactory.liveComment())
            }
        }).build()

        val vm = CommentsViewModel.ViewModel(
            env.toBuilder().currentUser(MockCurrentUser(currentUser)).build()
        )

        // Start the view model with an update.
        vm.intent(Intent().putExtra(IntentKey.UPDATE, UpdateFactory.update()))
        vm.outputs.insertComment().subscribe(commentSubscriber)

        // post a comment
        vm.inputs.postComment("Some Comment")

        commentSubscriber.assertValue(CommentFactory.liveComment())
    }
```

# 👀 See

https://user-images.githubusercontent.com/63934292/119713171-573e4e80-be59-11eb-8a74-83dba8549969.mp4


| Before 🐛 | After 🦋 |
| --- | --- |
|  |  |

# 📋 QA

- Go to your (any) backed project detail page
- Select ```comments```
- Type in a comment and ```POST```. You should see your comment added and scrolled to top

# Story 📖

https://kickstarter.atlassian.net/browse/NT-1914
